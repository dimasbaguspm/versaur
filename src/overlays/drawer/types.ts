/**
 * Drawer transition type options
 */
export type DrawerTransitionType = 'slide' | 'fade'
import type { ComponentPropsWithoutRef, HTMLAttributes, ReactNode } from 'react'
import type { VariantProps } from '@/utils/variants'
import type { drawerVariants } from './helpers'
import type { OverlayPortalProps } from '@/utils/overlay-portal'

/**
 * Base drawer variant props derived from the drawer variants
 */
export type DrawerVariantProps = VariantProps<typeof drawerVariants>

/**
 * Drawer positioning options
 */
export type DrawerPosition = 'left' | 'right'

/**
 * Drawer size options
 */
export type DrawerSize = 'sm' | 'md' | 'lg' | 'xl' | '3/4' | 'full'

/**
 * Drawer variant options
 */
export type DrawerVariant = 'default' | 'glass'

/**
 * Context value for the Drawer compound component
 */
export interface DrawerContextValue {
  /** Whether the drawer is open */
  isOpen: boolean
  /** Function to close the drawer */
  onClose: () => void
  /** Drawer position (left or right) */
  position: DrawerPosition
  /** Drawer size */
  size: DrawerSize
  /** Drawer variant */
  variant: DrawerVariant
  /** Drawer transition type */
  transitionType: DrawerTransitionType
}

/**
 * Props for the Drawer component (controlled component)
 */
export interface DrawerProps
  extends ComponentPropsWithoutRef<'div'>,
    OverlayPortalProps {
  /** Whether the drawer is open (required - controlled component) */
  isOpen: boolean
  /** Callback when the drawer should close */
  onClose: (open: boolean) => void
  /** Drawer position */
  position?: DrawerPosition
  /** Drawer size */
  size?: DrawerSize
  /** Drawer variant */
  variant?: DrawerVariant
  /** Drawer transition type */
  transitionType?: DrawerTransitionType
  /** Children components */
  children: ReactNode
}

/**
 * Props for the DrawerHeader component
 */
export interface DrawerHeaderProps extends ComponentPropsWithoutRef<'div'> {
  /** Children components */
  children: ReactNode
  hasTab?: boolean
}

export interface DrawerHeaderTabProps extends HTMLAttributes<HTMLDivElement> {
  /** Children components */
  children: ReactNode
}

/**
 * Props for the DrawerBody component
 */
export interface DrawerBodyProps extends ComponentPropsWithoutRef<'div'> {
  /** Children components */
  children: ReactNode
}

/**
 * Props for the DrawerFooter component
 */
export interface DrawerFooterProps extends ComponentPropsWithoutRef<'div'> {
  /** Children components */
  children: ReactNode
}

/**
 * Props for the DrawerOverlay component
 */
export type DrawerOverlayProps = ComponentPropsWithoutRef<'div'>
