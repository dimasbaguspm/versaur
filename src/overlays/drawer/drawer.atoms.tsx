import React, { forwardRef, type MouseEventHandler } from 'react'
import { XIcon } from 'lucide-react'
import { cn } from '@/utils/cn'
import { useDrawerContext } from './context'
import {
  drawerOverlayVariants,
  drawerHeaderVariants,
  drawerBodyVariants,
  drawerFooterVariants,
} from './helpers'
import type {
  DrawerHeaderProps,
  DrawerBodyProps,
  DrawerFooterProps,
  DrawerOverlayProps,
  DrawerHeaderTabProps,
} from './types'
import {
  ButtonIcon,
  Text,
  type ButtonIconProps,
  type TextProps,
} from '@/primitive'

/**
 * DrawerOverlay - Background overlay that appears behind the drawer
 * Always provides a dark blurred background to help users focus on the drawer content
 */
export const DrawerOverlay = React.forwardRef<
  HTMLDivElement,
  DrawerOverlayProps
>(({ className, ...props }, ref) => {
  const { isOpen, disableOverlayClickToClose, onClose } = useDrawerContext()

  const handleClick = () => {
    if (disableOverlayClickToClose) return
    onClose()
  }

  return (
    <div
      ref={ref}
      onClick={handleClick}
      className={cn(
        drawerOverlayVariants({
          state: isOpen ? 'open' : 'closed',
        }),
        className
      )}
      {...props}
    />
  )
})

/**
 * DrawerHeader - Header section of the drawer
 * Typically contains the title and close button
 */
export const DrawerHeader = React.forwardRef<HTMLDivElement, DrawerHeaderProps>(
  ({ children, className, hasTab, ...props }, ref) => {
    const { variant } = useDrawerContext()

    return (
      <div
        ref={ref}
        className={cn(
          drawerHeaderVariants({ variant, tab: Boolean(hasTab) }),
          className
        )}
        {...props}
      >
        {children}
      </div>
    )
  }
)

export const DrawerTitle = forwardRef<HTMLElement, TextProps>((props, ref) => {
  return <Text {...props} ref={ref} as='h3' fontSize='lg' />
})

export const DrawerCloseButton = forwardRef<
  HTMLButtonElement,
  Partial<ButtonIconProps>
>(({ onClick, ...props }, ref) => {
  const { onClose } = useDrawerContext()

  const handleOnClick: MouseEventHandler<HTMLButtonElement> = ev => {
    onClose()
    onClick?.(ev)
  }

  return (
    <ButtonIcon
      aria-label='Close drawer'
      {...props}
      onClick={handleOnClick}
      variant='ghost'
      size='sm'
      as={XIcon}
      ref={ref}
    />
  )
})

export const DrawerTab = forwardRef<HTMLDivElement, DrawerHeaderTabProps>(
  ({ children, className, ...props }, ref) => {
    return (
      <div ref={ref} className={cn('[&>div]:px-6', className)} {...props}>
        {children}
      </div>
    )
  }
)

/**
 * DrawerBody - Main content area of the drawer
 * Scrollable container for the drawer content
 */
export const DrawerBody = React.forwardRef<HTMLDivElement, DrawerBodyProps>(
  ({ children, className, ...props }, ref) => {
    const { variant } = useDrawerContext()

    return (
      <div
        ref={ref}
        className={cn(drawerBodyVariants({ variant }), className)}
        {...props}
      >
        {children}
      </div>
    )
  }
)

/**
 * DrawerFooter - Footer section of the drawer
 * Typically contains action buttons
 * Supports responsive flex behavior for optimal button layout across screen sizes
 */
export const DrawerFooter = React.forwardRef<HTMLDivElement, DrawerFooterProps>(
  ({ children, className, ...props }, ref) => {
    const { variant } = useDrawerContext()

    return (
      <div
        ref={ref}
        className={cn(drawerFooterVariants({ variant }), className)}
        {...props}
      >
        {children}
      </div>
    )
  }
)
