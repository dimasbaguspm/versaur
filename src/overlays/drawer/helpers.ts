import { cva } from '@/utils/variants'

/**
 * Drawer overlay variants using CVA
 * Always provides a dark blurred background to help users focus on the drawer content
 */
export const drawerOverlayVariants = cva(
  'fixed inset-0 z-50 transition-opacity duration-300 bg-foreground/30 backdrop-blur-md',
  {
    variants: {
      state: {
        open: 'opacity-100',
        closed: 'opacity-0 pointer-events-none',
      },
    },
    defaultVariants: {
      state: 'closed',
    },
  }
)

/**
 * Drawer content variants using CVA
 * Provides different sizes and positions for the drawer content
 */
export const drawerVariants = cva(
  'fixed z-50 shadow-xl flex flex-col max-w-full bg-white',
  {
    variants: {
      position: {
        left: 'left-0 top-0 bottom-0 border-r border-border',
        right: 'right-0 top-0 bottom-0 border-l border-border',
      },
      size: {
        sm: 'w-80',
        md: 'w-96',
        lg: 'w-[28rem]',
        xl: 'w-[32rem]',
        '3/4': 'w-[75vw]',
        full: 'w-full',
      },
      transitionType: {
        slide: 'transition-transform duration-300 ease-in-out',
        fade: 'transition-opacity duration-300 ease-in-out',
      },
    },
    defaultVariants: {
      position: 'right',
      size: 'md',
      transitionType: 'slide',
    },
  }
)

/**
 * Drawer header variants
 */
export const drawerHeaderVariants = cva(
  'flex justify-between items-center px-6 bg-white',
  {
    variants: {
      tab: {
        true: 'pt-4 pb-2',
        false: 'border-b border-border py-4',
      },
    },
    defaultVariants: {
      tab: false,
    },
  }
)

/**
 * Drawer body variants
 */
export const drawerBodyVariants = cva(
  'flex-1 overflow-y-auto px-6 py-4 bg-white'
)

/**
 * Drawer footer variants
 */
export const drawerFooterVariants = cva(
  'px-6 py-4 border-t border-border bg-white'
)
