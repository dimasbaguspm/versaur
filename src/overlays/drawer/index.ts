export { Drawer } from './drawer'
export type * from './types'
