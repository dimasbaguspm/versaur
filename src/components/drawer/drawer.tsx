import React, { useCallback, useEffect } from 'react'
import { cn } from '@/utils/cn'
import { DrawerContext } from './context'
import {
  DrawerOverlay,
  DrawerHeader,
  DrawerBody,
  DrawerFooter,
} from './drawer.atoms'
import { drawerVariants } from './helpers'
import type { DrawerProps, DrawerContextValue } from './types'

/**
 * Drawer - A controlled sliding drawer overlay component
 * Provides additional space for content with positioning, sizing, and glass variant options
 */
export const DrawerRoot: React.FC<DrawerProps> = ({
  children,
  isOpen,
  onClose,
  position = 'right',
  size = 'md',
  variant = 'default',
  className,
  ...props
}) => {
  const handleClose = useCallback(() => {
    onClose(false)
  }, [onClose])

  const contextValue: DrawerContextValue = {
    isOpen,
    onClose: handleClose,
    position,
    size,
    variant,
  }

  // Handle escape key to close drawer
  useEffect(() => {
    const handleEscape = (event: KeyboardEvent) => {
      if (event.key === 'Escape' && isOpen) {
        handleClose()
      }
    }

    if (isOpen) {
      document.addEventListener('keydown', handleEscape)
      document.body.style.overflow = 'hidden'
    }

    return () => {
      document.removeEventListener('keydown', handleEscape)
      document.body.style.overflow = 'unset'
    }
  }, [isOpen, handleClose])

  return (
    <DrawerContext.Provider value={contextValue}>
      <div
        className={cn(
          'fixed z-50 inset-0 pointer-events-none',
          isOpen && 'pointer-events-auto'
        )}
      >
        <DrawerOverlay />
        <div
          role={isOpen ? 'dialog' : undefined}
          aria-modal={isOpen ? 'true' : undefined}
          className={cn(
            drawerVariants({
              position,
              size,
              variant,
            }),
            // Apply closed transform when not open
            !isOpen && position === 'left' && '-translate-x-full',
            !isOpen && position === 'right' && 'translate-x-full',
            className
          )}
          {...props}
        >
          {isOpen && children}
        </div>
      </div>
    </DrawerContext.Provider>
  )
}

/**
 * Drawer - Compound component with sub-components attached
 * Provides a convenient API for using the drawer with all its parts
 */
export const Drawer = Object.assign(DrawerRoot, {
  Header: DrawerHeader,
  Body: DrawerBody,
  Footer: DrawerFooter,
})
