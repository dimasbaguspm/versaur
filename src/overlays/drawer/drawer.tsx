import React from 'react'
import { cn } from '@/utils/cn'
import { DrawerContext } from './context'
import {
  DrawerOverlay,
  DrawerHeader,
  DrawerBody,
  DrawerFooter,
  DrawerTab,
  DrawerCloseButton,
  DrawerTitle,
} from './drawer.atoms'
import { drawerVariants } from './helpers'
import { useEscapeClose } from '@/utils/use-escape-close'
import type { DrawerProps, DrawerContextValue } from './types'
import { OverlayPortal } from '@/utils/overlay-portal'

/**
 * Drawer - A controlled sliding drawer overlay component
 * Provides additional space for content with positioning, sizing, and glass variant options
 */
export const DrawerRoot: React.FC<DrawerProps> = ({
  container,
  children,
  isOpen,
  onClose,
  position = 'right',
  size = 'md',
  variant = 'default',
  transitionType = 'slide',
  disableOverlayClickToClose = false,
  disableEscapeKeyDown = false,
  className,
  ...props
}) => {
  const contextValue: DrawerContextValue = {
    isOpen,
    onClose,
    position,
    size,
    variant,
    transitionType,
    disableOverlayClickToClose,
    disableEscapeKeyDown,
  }

  const drawerRef = useEscapeClose(isOpen, onClose, disableEscapeKeyDown)

  return (
    <OverlayPortal container={container}>
      <DrawerContext.Provider value={contextValue}>
        <div
          className={cn(
            'fixed z-50 inset-0 pointer-events-none',
            isOpen && 'pointer-events-auto'
          )}
        >
          <DrawerOverlay />
          <div
            ref={drawerRef}
            tabIndex={-1}
            role={isOpen ? 'dialog' : undefined}
            aria-modal={isOpen ? 'true' : undefined}
            className={cn(
              drawerVariants({
                position,
                size,
                variant,
                transitionType,
              }),
              transitionType === 'slide'
                ? [
                    !isOpen && position === 'left' && '-translate-x-full',
                    !isOpen && position === 'right' && 'translate-x-full',
                  ]
                : [
                    'left-0 right-0 top-0 bottom-0',
                    isOpen ? 'opacity-100' : 'opacity-0 pointer-events-none',
                  ],
              className
            )}
            {...props}
          >
            {isOpen && children}
          </div>
        </div>
      </DrawerContext.Provider>
    </OverlayPortal>
  )
}

/**
 * Drawer - Compound component with sub-components attached
 * Provides a convenient API for using the drawer with all its parts
 */
export const Drawer = Object.assign(DrawerRoot, {
  Header: DrawerHeader,
  Title: DrawerTitle,
  CloseButton: DrawerCloseButton,
  Tab: DrawerTab,
  Body: DrawerBody,
  Footer: DrawerFooter,
})
