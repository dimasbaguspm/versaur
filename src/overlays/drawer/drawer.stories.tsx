import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import { Button } from '@/primitive/button'
import { Drawer } from './drawer'
import type { DrawerProps } from './types'
import { X, Bell } from 'lucide-react'

// Wrapper component to handle state in stories
const DrawerWrapper = ({
  children,
  triggerLabel = 'Open Drawer',
  ...drawerProps
}: Partial<DrawerProps> & {
  children: React.ReactNode
  triggerLabel?: string
}) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <>
      <Button onClick={() => setIsOpen(true)}>{triggerLabel}</Button>
      <Drawer {...drawerProps} isOpen={isOpen} onClose={() => setIsOpen(false)}>
        {children}
      </Drawer>
    </>
  )
}

const meta: Meta<typeof Drawer> = {
  title: 'Overlays/Drawer',
  component: Drawer,
  parameters: {
    docs: {
      description: {
        component:
          'A controlled sliding drawer overlay component that provides additional space for content. Supports left/right positioning, multiple sizes, and glass variant where the drawer content becomes transparent with backdrop blur effects while the overlay remains dark for focus.',
      },
    },
  },
  argTypes: {
    isOpen: {
      control: 'boolean',
      description: 'Whether the drawer is open (controlled)',
    },
    position: {
      control: 'select',
      options: ['left', 'right'],
      description: 'Position where the drawer slides in from',
    },
    size: {
      control: 'select',
      options: ['sm', 'md', 'lg', 'xl', '3/4', 'full'],
      description: 'Size of the drawer',
    },
    variant: {
      control: 'select',
      options: ['default', 'glass'],
      description:
        'Visual variant of the drawer. Glass makes the drawer content transparent with backdrop blur',
    },
  },
}

export default meta
type Story = StoryObj<typeof Drawer>

/**
 * Default drawer with standard settings
 * Shows a basic controlled drawer with header, body, and footer sections
 */
export const Default: Story = {
  render: args => (
    <DrawerWrapper {...args}>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Drawer Title</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <p className='text-gray-600 mb-4'>
          This is the main content area of the drawer. You can put any content
          here.
        </p>
        <p className='text-gray-600'>
          The drawer supports different sizes, positions, and variants to match
          your design needs.
        </p>
      </Drawer.Body>
      <Drawer.Footer>
        <Button variant='ghost'>Cancel</Button>
        <Button>Save Changes</Button>
      </Drawer.Footer>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'md',
    variant: 'default',
  },
}

/**
 * Left position drawer
 * Demonstrates a drawer that slides in from the left side
 */
export const LeftPosition: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Settings'>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Settings</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <div className='space-y-4'>
          <div>
            <h4 className='font-medium mb-2'>Appearance</h4>
            <p className='text-sm text-gray-600'>Customize your experience</p>
          </div>
          <div>
            <h4 className='font-medium mb-2'>Notifications</h4>
            <p className='text-sm text-gray-600'>
              Manage notification preferences
            </p>
          </div>
          <div>
            <h4 className='font-medium mb-2'>Privacy</h4>
            <p className='text-sm text-gray-600'>
              Control your privacy settings
            </p>
          </div>
        </div>
      </Drawer.Body>
    </DrawerWrapper>
  ),
  args: {
    position: 'left',
    size: 'md',
    variant: 'default',
  },
}

/**
 * Glass variant drawer
 * Shows the glass variant with transparent backdrop blur effects on the drawer content
 * The overlay remains dark to maintain focus, while the drawer itself becomes glass-like
 */
export const GlassVariant: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Open Glass Drawer' size='full'>
      <Drawer.Body>
        <p className='text-white mb-4'>
          This drawer uses the glass variant where the drawer content itself
          becomes transparent with backdrop blur effects.
        </p>
        <p className='text-white mb-4'>
          The overlay remains dark to help you focus on the drawer content,
          while the drawer background creates a frosted glass appearance.
        </p>
        <p className='text-white'>
          Perfect for modern interfaces that need subtle, elegant overlay
          effects while maintaining visual connection with the underlying
          content.
        </p>
      </Drawer.Body>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'lg',
    variant: 'glass',
  },
}

/**
 * Large size drawer
 * Demonstrates different drawer sizes
 */
export const LargeSize: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Open Large Drawer'>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Notifications</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <div className='space-y-4'>
          {Array.from({ length: 8 }, (_, i) => (
            <div key={i} className='p-4 border border-gray-200 rounded-lg'>
              <div className='flex items-start justify-between'>
                <div>
                  <h4 className='font-medium'>Notification {i + 1}</h4>
                  <p className='text-sm text-gray-600 mt-1'>
                    This is a sample notification message with more details
                    about the event.
                  </p>
                  <span className='text-xs text-gray-400 mt-2 block'>
                    {new Date(Date.now() - i * 3600000).toLocaleString()}
                  </span>
                </div>
                <Bell size={16} className='text-blue-500 mt-1' />
              </div>
            </div>
          ))}
        </div>
      </Drawer.Body>
      <Drawer.Footer>
        <Button variant='ghost'>Mark all as read</Button>
        <Button>Close</Button>
      </Drawer.Footer>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'xl',
    variant: 'default',
  },
}

/**
 * Three-quarters width drawer
 * Shows a drawer that takes 75% of the viewport width, ideal for dashboards or complex forms
 */
export const ThreeQuartersWidth: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Open 3/4 Width'>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Dashboard View</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <div className='space-y-6'>
          <div className='grid grid-cols-1 lg:grid-cols-2 gap-4'>
            <div className='p-4 border border-gray-200 rounded-lg'>
              <h4 className='font-medium mb-2'>Analytics Overview</h4>
              <p className='text-gray-600 text-sm'>
                The 3/4 width drawer provides substantial space for complex
                layouts like dashboards, detailed forms, or content editing
                interfaces.
              </p>
            </div>
            <div className='p-4 border border-gray-200 rounded-lg'>
              <h4 className='font-medium mb-2'>Performance Metrics</h4>
              <p className='text-gray-600 text-sm'>
                This size strikes a balance between providing ample working
                space while maintaining some background context visibility.
              </p>
            </div>
          </div>
          <div className='p-4 border border-gray-200 rounded-lg'>
            <h4 className='font-medium mb-4'>Recent Activity</h4>
            <div className='space-y-3'>
              {Array.from({ length: 5 }, (_, i) => (
                <div
                  key={i}
                  className='flex items-center justify-between py-2 border-b border-gray-100 last:border-b-0'
                >
                  <div>
                    <p className='font-medium text-sm'>Activity {i + 1}</p>
                    <p className='text-xs text-gray-500'>
                      Description of the activity item
                    </p>
                  </div>
                  <span className='text-xs text-gray-400'>
                    {new Date(Date.now() - i * 1800000).toLocaleTimeString()}
                  </span>
                </div>
              ))}
            </div>
          </div>
        </div>
      </Drawer.Body>
      <Drawer.Footer>
        <Button variant='ghost'>Reset</Button>
        <Button>Apply Changes</Button>
      </Drawer.Footer>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: '3/4',
    variant: 'default',
  },
}

/**
 * Full width drawer
 * Shows a drawer that takes the full width of the viewport
 */
export const FullWidth: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Open Full Width'>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Full Width Drawer</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <div className='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6'>
          {Array.from({ length: 6 }, (_, i) => (
            <div key={i} className='p-6 border border-gray-200 rounded-lg'>
              <h4 className='font-medium mb-2'>Card {i + 1}</h4>
              <p className='text-gray-600'>
                This is a sample card in the full-width drawer layout.
              </p>
            </div>
          ))}
        </div>
      </Drawer.Body>
      <Drawer.Footer>
        <Button variant='ghost'>Cancel</Button>
        <Button>Save All</Button>
      </Drawer.Footer>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'full',
    variant: 'default',
  },
}

/**
 * Custom footer layout (non-responsive)
 * Shows how to disable the responsive behavior for custom button arrangements
 */
export const CustomFooterLayout: Story = {
  render: args => (
    <DrawerWrapper {...args} triggerLabel='Custom Footer'>
      <Drawer.Header>
        <div className='flex items-center justify-between'>
          <h3 className='text-lg font-semibold'>Custom Layout</h3>
          <Button variant='ghost' size='sm'>
            <X size={16} />
          </Button>
        </div>
      </Drawer.Header>
      <Drawer.Body>
        <p className='text-gray-600 mb-4'>
          This drawer footer has custom styling with responsive behavior
          disabled. The buttons maintain their custom layout across all screen
          sizes.
        </p>
        <p className='text-gray-600'>
          Use `responsiveFlex={false}` when you need complete control over
          footer layout.
        </p>
      </Drawer.Body>
      <Drawer.Footer responsiveFlex={false}>
        <div className='flex items-center justify-between w-full'>
          <Button variant='ghost' size='sm'>
            <Bell size={16} className='mr-2' />
            Notify
          </Button>
          <div className='flex gap-2'>
            <Button variant='ghost'>Cancel</Button>
            <Button>Apply</Button>
          </div>
        </div>
      </Drawer.Footer>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'md',
    variant: 'default',
  },
  parameters: {
    docs: {
      description: {
        story:
          'When `responsiveFlex={false}`, the footer does not apply responsive behavior. You have full control over the layout and can implement custom arrangements that remain consistent across all screen sizes.',
      },
    },
  },
}

export const Fade: Story = {
  render: args => (
    <DrawerWrapper
      {...args}
      transitionType='fade'
      triggerLabel='Open Fade Drawer'
    >
      <Drawer.Header>
        <h3 className='text-lg font-semibold'>Fade Transition</h3>
      </Drawer.Header>
      <Drawer.Body>
        <p>This drawer uses a fade transition instead of sliding in/out.</p>
      </Drawer.Body>
    </DrawerWrapper>
  ),
  args: {
    position: 'right',
    size: 'full',
    variant: 'default',
    transitionType: 'fade',
  },
}
