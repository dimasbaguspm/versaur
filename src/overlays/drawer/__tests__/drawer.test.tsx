import { describe, it, expect } from 'vitest'
import { render, screen, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../drawer.stories'

const { Default, LeftPosition, LargeSize, FullWidth, Fade } =
  composeStories(stories)

describe('Drawer', () => {
  describe('Default Story', () => {
    it('renders the drawer trigger button', () => {
      render(<Default />)
      expect(screen.getByText('Open Drawer')).toBeInTheDocument()
    })

    it('opens the drawer when trigger is clicked', () => {
      render(<Default />)

      // Initially, drawer content should not be visible
      expect(screen.queryByText('Drawer Title')).not.toBeInTheDocument()

      // Click the trigger button
      fireEvent.click(screen.getByText('Open Drawer'))

      // Drawer content should now be visible
      expect(screen.getByText('Drawer Title')).toBeInTheDocument()
      expect(
        screen.getByText(
          'This is the main content area of the drawer. You can put any content here.'
        )
      ).toBeInTheDocument()
    })

    it('closes the drawer when escape key is pressed', () => {
      render(<Default />)

      // Open the drawer
      fireEvent.click(screen.getByText('Open Drawer'))
      expect(screen.getByText('Drawer Title')).toBeInTheDocument()

      // Press escape key
      fireEvent.keyDown(document, { key: 'Escape', code: 'Escape' })

      // Drawer content should be hidden
      expect(screen.queryByText('Drawer Title')).not.toBeInTheDocument()
    })
  })

  describe('Left Position Story', () => {
    it('renders the settings drawer with left position', () => {
      render(<LeftPosition />)

      fireEvent.click(screen.getByText('Settings'))

      expect(screen.getByText('Appearance')).toBeInTheDocument()
      expect(screen.getByText('Notifications')).toBeInTheDocument()
      expect(screen.getByText('Privacy')).toBeInTheDocument()
    })
  })

  describe('Large Size Story', () => {
    it('renders the large drawer with notifications', () => {
      render(<LargeSize />)

      // Click the trigger button
      fireEvent.click(screen.getByText('Open Large Drawer'))

      // Check that the drawer content is visible
      expect(screen.getByText('Notifications')).toBeInTheDocument()
      expect(screen.getByText('Notification 1')).toBeInTheDocument()
      expect(screen.getByText('Mark all as read')).toBeInTheDocument()
    })
  })

  describe('Full Width Story', () => {
    it('renders the full width drawer', () => {
      render(<FullWidth />)

      // Click the trigger button
      fireEvent.click(screen.getByText('Open Full Width'))

      // Check that the drawer content is visible
      expect(screen.getByText('Full Width Drawer')).toBeInTheDocument()
      expect(screen.getByText('Card 1')).toBeInTheDocument()
      expect(screen.getByText('Save All')).toBeInTheDocument()
    })
  })

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      render(<Default />)

      // Open the drawer
      fireEvent.click(screen.getByText('Open Drawer'))

      // Check ARIA attributes
      const dialog = screen.getByRole('dialog')
      expect(dialog).toHaveAttribute('aria-modal', 'true')
    })

    it('traps focus within the drawer when open', () => {
      render(<Default />)

      // Open the drawer
      fireEvent.click(screen.getByText('Open Drawer'))

      // The drawer content should be in the document
      expect(screen.getByRole('dialog')).toBeInTheDocument()
    })
  })

  describe('Overlay Click', () => {
    it('closes drawer when overlay is clicked', () => {
      render(<Default />)

      // Open the drawer
      fireEvent.click(screen.getByText('Open Drawer'))
      expect(screen.getByText('Drawer Title')).toBeInTheDocument()

      // Click the overlay (the overlay should be in the DOM when drawer is open)
      const overlay = document.querySelector('[class*="drawerOverlayVariants"]')
      if (overlay) {
        fireEvent.click(overlay)
        // Drawer should close
        expect(screen.queryByText('Drawer Title')).not.toBeInTheDocument()
      }
    })
  })

  describe('Fade Transition Story', () => {
    it('renders the fade transition drawer', () => {
      render(<Fade />)

      // Click the trigger button
      fireEvent.click(screen.getByText('Open Fade Drawer'))

      // Check that the drawer content is visible
      expect(screen.getByText('Fade Transition')).toBeInTheDocument()
      expect(
        screen.getByText(
          'This drawer uses a fade transition instead of sliding in/out.'
        )
      ).toBeInTheDocument()
    })
  })
})
