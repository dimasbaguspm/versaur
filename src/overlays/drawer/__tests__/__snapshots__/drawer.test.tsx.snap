// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer > Default Story > matches snapshot when open 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
    class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 px-4 text-sm min-w-[2.5rem]"
    type="button"
  >
    Open Drawer
  </button>
  <div
    class="fixed z-50 inset-0 pointer-events-auto"
  >
    <div
      class="fixed inset-0 z-50 transition-opacity duration-300 bg-black/30 backdrop-blur-md opacity-100"
    />
    <div
      aria-modal="true"
      class="fixed z-50 shadow-xl flex flex-col right-0 top-0 bottom-0 border-l w-96 bg-white border-gray-200 transition-transform duration-300 ease-in-out"
      role="dialog"
    >
      <div
        class="flex-shrink-0 px-6 bg-white border-b border-border py-4"
      >
        <div
          class="flex items-center justify-between"
        >
          <h3
            class="text-lg font-semibold"
          >
            Drawer Title
          </h3>
          <button
            aria-disabled="false"
            aria-label="Close drawer"
            class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-white text-foreground hover:bg-ghost-soft focus-visible:ring-ghost-soft focus-visible:ring-offset-white h-8 w-8 rounded-md"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-x inline-flex items-center justify-center w-4 h-4 text-ghost"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex-1 overflow-y-auto px-6 py-4 bg-white"
      >
        <p
          class="text-gray-600 mb-4"
        >
          This is the main content area of the drawer. You can put any content here.
        </p>
        <p
          class="text-gray-600"
        >
          The drawer supports different sizes, positions, and variants to match your design needs.
        </p>
      </div>
      <div
        class="flex-shrink-0 px-6 py-4 border-t bg-white border-gray-200 flex flex-row gap-3 sm:justify-end [&>*]:sm:flex-grow-0 [&>*]:flex-grow"
      >
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-white text-foreground hover:bg-ghost-soft focus-visible:ring-ghost-soft focus-visible:ring-offset-white h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          Cancel
        </button>
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          Save Changes
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
